mindmeld/augmentation.py

```diff
@@ -241,7 +302,160 @@ def __init__(self, batch_size, language, paths, path_suffix, resource_loader):
             "max_length": 60,
         }
 
-    def _generate_paraphrases(self, queries):
+    def _download_model(self):
+        logger.info("Downloading paraphrase model from %s", PARAPHRASER_RETAIN_ENTITIES_URL)
+
+        # Make the folder that will contain the model folder
+        if not os.path.exists(EMBEDDINGS_FOLDER_PATH):
+            os.makedirs(EMBEDDINGS_FOLDER_PATH)
+
+        with TqdmUpTo(unit="B", unit_scale=True, miniters=1, desc='') as t:
+            try:
+                urlretrieve(PARAPHRASER_RETAIN_ENTITIES_URL,
+                            PARAPHRASER_FILE_PATH,
+                            reporthook=t.update_to)
+            except ConnectionError as e:
+                logger.error("Model download failed with error: %s", e)
+                return
+        try:
+            with zipfile.ZipFile(PARAPHRASER_FILE_PATH, 'r') as zip_ref:
+                zip_ref.extractall(EMBEDDINGS_FOLDER_PATH)
+            os.remove(PARAPHRASER_FILE_PATH)
+        except zipfile.BadZipfile:
+            logger.error("Unable to extract zip file. Try downloading the model again.")
+
+    def _prepare_inputs(self, processed_queries):
+        """Processes input as expected by the two different English models
+        Example:
+            The default model requires just <unannotated text>
+            The retain_entities model requires <unannotated text> <EOS> <entity values>
+
+        Args:
+            processed queries (list(ProcessedQuery)): List of ProcessedQuery objects from the app
+
+        Return:
+            model_inputs (list(str)): List of queries to be paraphrased in the
+                                      format required by the models
+            processed_queries (list(ProcessedQuery)): List of ProcessedQuery
+                                                       with entity annotations
+        """
+        model_inputs = []
+        for processed_query in processed_queries:
+            processed_query_text = processed_query.query.text.strip()
+            if self.retain_entities:
+                text = [processed_query_text.lower(), EOS_TOKEN]
+                for entity in processed_query.entities:
```

Comment:
does it matter what entity comes first (I assume we sort by span start index)?

Reply:
I sort the list when replacing the entities